_source/_posts/2023-07-27-spring-graphql-react.md

```diff
@@ -0,0 +1,1794 @@
+  backendAPI.interceptors.request.eject(requestInterceptor);
+  backendAPI.interceptors.response.eject(responseInterceptor);
+};
+
+export const setInterceptors = (accessToken: String) => {
+
+  clearInterceptors();
+
+  requestInterceptor = backendAPI.interceptors.request.use(
+    // @ts-expect-error
+    function (config) {
+      return {
+        ...config,
+        headers: {
+          ...config.headers,
+          Authorization: `Bearer ${accessToken}`,
+        },
+      };
+    },
+    function (error) {
+      console.log("request interceptor error", error);
+      return Promise.reject(error);
+    }
+  );
+};
+```
+
+Add the file `src/hooks/useAccessToken.tsx` with the following content:
+
+__useAccessToken.tsx__
+```tsx
+import { setInterceptors } from "@/services/auth";
+import { useAuth0 } from "@auth0/auth0-react";
+import { useCallback, useState } from "react";
+
+export const useAccessToken = () => {
+  const { isAuthenticated, getAccessTokenSilently } = useAuth0();
+  const [accessToken, setAccessToken] = useState("");
+
+  const saveAccessToken = useCallback(async () => {
+    if (isAuthenticated) {
+      try {
+        const tokenValue = await getAccessTokenSilently();
+        if (accessToken !== tokenValue) {
+          setInterceptors(tokenValue);
+          setAccessToken(tokenValue);
+        }
+      } catch (err) {
+        // Inactivity timeout
+        console.log("getAccessTokenSilently error", err);
+      }
+    }
+  }, [getAccessTokenSilently, isAuthenticated, accessToken]);
+
+  return {
+    saveAccessToken,
+  };
+};
+```
+
+The hook will call Auth0 `getAccessTokenSilently()` and will trigger a token refresh if the access token is expired. Then it will update axios interceptors to set the updated bearer token value in the request headers.
+Create the `useAsyncWithToken` hook:
+
+__useAsyncWithToken.tsx__
+```tsx
+import { useAccessToken } from "./useAccessToken";
+import { useAsync } from "react-use-custom-hooks";
+
+export const useAsyncWithToken = <T, P, E = string>(
+  asyncOperation: () => Promise<T>, deps: any[]
+) => {
+  const { saveAccessToken } = useAccessToken();
+  const [ data, loading, error ] = useAsync(async () => {
+    await saveAccessToken();
+    return asyncOperation();
+  }, {},  deps);
+
+  return {
+    data,
+    loading,
+    error
+  };
+};
+```
+
+Update the calls in the `CompanyTableContainer` component to use the `useAsyncWithToken` hook:
+
+__CompanyTableContainer.tsx__
+```tsx
+...
+const {
+  data: dataList,
+  loading: loadingList,
+  error: errorList,
+} = useAsyncWithToken(
+  () => CompanyApi.getCompanyList({ page: page - 1}),
+  [props.page]
+);
+
+const { data: dataCount } = useAsyncWithToken(
+  () => CompanyApi.getCompanyCount(),
+  []
+);
+...
+```
+
+Run the application with:
+
+```shell
+npm run dev
+```
+
+Go to http://localhost:3000 and you should be redirected to the Auth0 universal login page. After login in, you should see the companies list again.
```

Comment:
```suggestion
Go to `http://localhost:3000` and you should be redirected to the Auth0 universal login page. After logging in, you should see the companies list again.
```

Reply:
This took quite some time to load, around 10 seconds. Is that expected?

Reply:
I also saw this error in my console:

```
FetchError: request to https://fonts.gstatic.com/s/ubuntu/v20/4iCs6KVjbNBYlgoKcQ72nU6AF7xm.woff2 failed, reason: getaddrinfo ENOTFOUND fonts.gstatic.com
    at ClientRequest.<anonymous> (/Users/mraible/Downloads/react-graphql/node_modules/next/dist/compiled/node-fetch/index.js:1:65756)
    at ClientRequest.emit (node:events:513:28)
    at TLSSocket.socketErrorListener (node:_http_client:502:9)
    at TLSSocket.emit (node:events:513:28)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  type: 'system',
  errno: 'ENOTFOUND',
  code: 'ENOTFOUND'
}
- error Failed to download `Ubuntu` from Google Fonts. Using fallback font instead.
```

Reply:
After redirecting to `/callback`, I waited a minute before I gave up and refreshed the page. I opened a new incognito window and tried again. This time it took 20 seconds to load and redirect to the login page. Why is it so slow every time?

Reply:
When it came back to `/callback`, I waited for 5 minutes before I gave up. It just spins.

<img width="1176" alt="Screenshot 2023-09-08 at 11 52 25 AM" src="https://github.com/oktadev/okta-blog/assets/17892/d138b46c-686e-48a9-b0bf-81918bf46b5c">

Reply:
> ENOTFOUND

I also had issues and thought it was my networking. It seems there is some issue with [next/font](https://github.com/vercel/next.js/issues/45080). Will try a workaround.

Reply:
> I also saw this error in my console:
> 
> ```
> FetchError: request to https://fonts.gstatic.com/s/ubuntu/v20/4iCs6KVjbNBYlgoKcQ72nU6AF7xm.woff2 failed, reason: getaddrinfo ENOTFOUND fonts.gstatic.com
>     at ClientRequest.<anonymous> (/Users/mraible/Downloads/react-graphql/node_modules/next/dist/compiled/node-fetch/index.js:1:65756)
>     at ClientRequest.emit (node:events:513:28)
>     at TLSSocket.socketErrorListener (node:_http_client:502:9)
>     at TLSSocket.emit (node:events:513:28)
>     at emitErrorNT (node:internal/streams/destroy:151:8)
>     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
>     at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
>   type: 'system',
>   errno: 'ENOTFOUND',
>   code: 'ENOTFOUND'
> }
> - error Failed to download `Ubuntu` from Google Fonts. Using fallback font instead.
> ```

I also had issues and thought it was my networking. It seems there is some issue with [next/font](https://github.com/vercel/next.js/issues/45080). Will try a workaround.